interp/numTok.ml

```diff
@@ -1,52 +1,219 @@
+         | _ -> ""
+         end
+      | _ -> "" in
+    { int = i; frac = f; exp = e }
+
+  let to_string n =
+    n.int ^ (if n.frac = "" then "" else "." ^ n.frac) ^ n.exp
+
+  let of_string s =
+    if s = "" || s.[0] < '0' || s.[0] > '9' then None else
+      let strm = Stream.of_string (s ^ "  ") in
+      let n = parse strm in
+      if Stream.count strm >= String.length s then Some n else None
+
+  let to_nat ?loc = function
+    | { int = i; frac = ""; exp = "" } -> i
+    | _ -> CErrors.user_err ?loc (Pp.str "This number is not a natural number.")
+
+  let is_nat = function
+    | { int = _; frac = ""; exp = "" } -> true
+    | _ -> false
+
+end
+
+open Unsigned
+
+module Signed =
+struct
+
+  type t = sign * Unsigned.t
+
+  let equal (s1,n1) (s2,n2) =
+    s1 = s2 && equal n1 n2
+
+  let is_zero s =
+    let rec aux i =
+      Int.equal (String.length s) i || (s.[i] == '0' && aux (i+1))
+    in aux 0
+  let is_zero = function
+    | (SPlus,{int;frac;exp}) -> is_zero int && is_zero frac
+    | _ -> false
+
+  let of_decimal_and_exponent (sign,int) f e =
+    let exp = match e with Some e -> "e" ^ SignedNat.to_string e | None -> "" in
+    let frac = match f with Some f -> UnsignedNat.to_string f | None -> "" in
+    sign, { int; frac; exp }
+
+  let to_decimal_and_exponent (sign, { int; frac; exp }) =
+    let e =
+      if exp = "" then None else
+        Some (match exp.[1] with
+        | '-' -> SMinus, String.sub exp 2 (String.length exp - 2)
+        | '+' -> SPlus, String.sub exp 2 (String.length exp - 2)
+        | _ -> SPlus, String.sub exp 1 (String.length exp - 1)) in
+    let f = if frac = "" then None else Some frac in
+    (sign, int), f, e
+
+  let of_int_and_exponent i e =
+    of_decimal_and_exponent i None e
+
+  let of_nat i =
+    (SPlus,{ int = i; frac = ""; exp = "" })
+
+  let of_int (s,i) =
+    (s,{ int = i; frac = ""; exp = "" })
+
+  let of_int_string s = of_int (SignedNat.of_string s)
+
+  let to_string (sign,u) =
+    (match sign with SPlus -> "" | SMinus -> "-") ^ Unsigned.to_string u
+
+  let to_int = function
+    | (s, { int = i; frac = ""; exp = "" }) -> Some (s,i)
+    | _ -> None
+
+  let is_int n = match to_int n with None -> false | Some _ -> true
+
+  let of_string s =
+    if s = "" || s = "-" ||
+       (s.[0] < '0' || s.[0] > '9') && (s.[0] <> '-' || s.[1] < '0' || s.[1] > '9') then None else
+      let strm = Stream.of_string (s ^ "  ") in
+      let sign = if s.[0] = '-' then (Stream.junk strm; SMinus) else SPlus in
+      let n = parse strm in
+      if Stream.count strm >= String.length s then Some (sign,n) else None
+
+  let to_bigint = function
+    | (sign, { int = n; frac = ""; exp = "" }) ->
+      let n = String.(concat "" (split_on_char '_' n)) in
+      Some (SignedNat.to_bigint (sign,n))
+    | _ -> None
+
+  let of_bigint n =
+    let sign, int = SignedNat.of_bigint n in
+    (sign, { int; frac = ""; exp = "" })
+
+  let to_bigint_fraction (s, { int; frac; exp }) =
```

Comment:
Don't we need to remove the underscores (like in `to_bigint` just above)?

Reply:
Yes, well spotted.

I'm going tonight to clarify the semantics of the different `to_string` functions.